Rpi5: mmc add adma support

GrygiriiS opened the pull request:
This PR enables ADMA2 support for rpi5 SDHC controller.
The ADMA2 is enabled by default.

Commit:
drivers: sdhc: sdhci_common: add adma2 support

This patch adds SDHC ADMA2 support (32-bits only). The consumer expected to
provide preallocated table of ADMA2 descriptors.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Acked-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

Commit:
drivers: sdhc: brcm_bcm2712: enable adma2

Add support and enable SDHC ADMA2 by default.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Acked-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

GrygiriiS commented:
@oleksiimoisieiev @dsemenets  @lorc 
cc: @xakep-amatop

dsemenets commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -453,7 +461,18 @@ static int bcm2712_sdhci_init(const struct device *dev)
+#if defined(CONFIG_BRCM_BCM2712_XFER_ADMA)
+#define BRCM_BCM2712_ADMA_DESCS(inst)                                                              \
+	static struct sdhc_adma_desc                                                               \
+		adma_descs_##inst[CONFIG_BRCM_BCM2712_ADMA_DESCS_NUM] __nocache;
```
I would prefer something like this:
```
#ifdef CONFIG_NOCACHE_MEMORY
#define __NOCACHE __attribute__((__section__(".nocache")))
#else
#define __NOCACHE
#endif

static struct sdhc_adma_desc adma_descs_##inst[CONFIG_BRCM_BCM2712_ADMA_DESCS_NUM] __NOCACHE;
```

GrygiriiS replied:
It's already done in this way in Zephyr core, see `section_tags.h`:
```
#if defined(CONFIG_NOCACHE_MEMORY)
#define __nocache __in_section_unique(_NOCACHE_SECTION_NAME)
#define __nocache_noinit __nocache
#else
#define __nocache
#define __nocache_noinit __noinit
#endif /* CONFIG_NOCACHE_MEMORY */
```

dsemenets commented:
`Acked-by: Dmytro Semenets <dmytro_semenets@epam.com>`

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+		cmd |= SDHC_ADMA_DESC_END;
+	}
+
+	desc->cmd = cmd;
```
As far as I can see, this pertains to a hardware structure, and you're populating a structure in a shared file that needs to be usable across different devices. Should you fix descriptor filling for both little-endian and big-endian machines?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -339,10 +341,16 @@ static int bcm2712_sdhci_init_host_props(const struct device *dev)
+		if (ret) {
+			LOG_DEV_ERR(dev, "enable adma failed (%d)", ret);
+		}
+	} else if (IS_ENABLED(CONFIG_BRCM_BCM2712_XFER_SDMA)) {
```
should you add a check of `host_caps->`[sdma_support](https://elixir.bootlin.com/zephyr/v3.7.0-rc1/C/ident/sdma_support) here?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+
+#if !defined(CONFIG_NOCACHE_MEMORY)
+	sys_cache_data_flush_range(sdhci_ctx->adma_descs,
+				   (sdhci_ctx->adma_descs_num * sizeof(struct sdhc_adma_desc)));
```
Should we flush only the descriptions that were written in this function?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+		desc++;
+	}
+
+	sdhc_adma_desc_fill(desc, dma_addr, trans_bytes, true);
```
If you have `desc_count` overflow, `trans_bytes` can be several times larger than `SDHC_ADMA_DESC_MAX_LEN`.

GrygiriiS replied:
fixed. Added sdhci_dma_check() to verify DMA xfer

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+				   (sdhci_ctx->adma_descs_num * sizeof(struct sdhc_adma_desc)));
+#endif
+
+	sys_write32((uintptr_t)sdhci_ctx->adma_descs, sdhci_ctx->reg_base + SDHCI_ADMA_ADDRESS);
```
you can use `POINTER_TO_UINT` here instead of `uintptr_t`

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+	}
+
+	while (--desc_count) {
+		sdhc_adma_desc_fill(desc, dma_addr, SDHC_ADMA_DESC_MAX_LEN, false);
```
So, if there is an overflow in the number of descriptors (check above), we will have memory corruption here because you didn't limit it to `adma_descs_num`.

-----
Even if you restrict `desc_count` to the maximum value, you still have problems in one of the caller functions. Look at `sdhci_prepare_data`:
```
sys_write16(data->blocks, sdhci_ctx->reg_base + SDHCI_BLOCK_COUNT);
```

and of course, there will be remaining data for transfer.

The SD Host Controller standard has an interesting comment about block size, this one:
```
Total Length = Length 1 + Length 2 + Length 3 + ... + Length n
= multiple of Block Size

If total length of a descriptor were not multiple of block size, ADMA2 transfer might not be terminated. In
this case, the transfer should be aborted by data timeout.
```
maybe in this case if timeout error exists it is needed to retriger transwer in non-DMA mode or SDMA mode


and another one comment:
```
Block Count register is defined as 16-bit register and it limits the maximum of 65535 blocks transfer. If
ADMA2 operation is less than or equal 65535 blocks transfer, Block Count register can be used. In this
case, total length of Descriptor Table shall be equivalent to multiply block size and block count. If ADMA2
operation is more than 65535 blocks transfer, Block Count register shall be disabled by setting 0 to Block
Count Enable in the Transfer Mode Register. In this case, length of data transfer is not designated by
block count but Descriptor Table
```

GrygiriiS replied:
ok. Got it missed proper error handling,

GrygiriiS replied:
fixed. Added sdhci_dma_check() to verify DMA xfer

xakep-amatop replied:
pls, fix this part of comment too:
```
Block Count register is defined as 16-bit register and it limits the maximum of 65535 blocks transfer. If
ADMA2 operation is less than or equal 65535 blocks transfer, Block Count register can be used. In this
case, total length of Descriptor Table shall be equivalent to multiply block size and block count. If ADMA2
operation is more than 65535 blocks transfer, Block Count register shall be disabled by setting 0 to Block
Count Enable in the Transfer Mode Register. In this case, length of data transfer is not designated by
block count but Descriptor Table
```

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+	uint32_t desc_count = DIV_ROUND_UP(trans_bytes, SDHC_ADMA_DESC_MAX_LEN);
+	struct sdhc_adma_desc *desc = sdhci_ctx->adma_descs;
+
+	if (desc_count >= sdhci_ctx->adma_descs_num) {
```
I believe we need to fix it. It won't be easy, but it's possible.

I have another solution for you, a lighter one: don't choose between `SDMA` or `ADMA` at the start. Use both of them, and in case you don't have enough `ADMA` descriptors, `SDMA` will be chosen by default at runtime (as far as I see, there isn't such problems with it).

GrygiriiS replied:
Sorry, I do not understand why is it needed and what is the problem? The Linux working this way.

xakep-amatop replied:
I reviewed the latest version of the Linux kernel a little bit, and it seems like they dynamically allocate ADMA descriptors.
I did it a few days ago, so I might be mistaken. If you want, I can check it again.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+	uint32_t desc_count = DIV_ROUND_UP(trans_bytes, SDHC_ADMA_DESC_MAX_LEN);
+	struct sdhc_adma_desc *desc = sdhci_ctx->adma_descs;
+
+	if (desc_count >= sdhci_ctx->adma_descs_num) {
```
Is it a problem if `desc_count` is equal to `adma_descs_num`?

GrygiriiS replied:
No. Transfer will fail and desc table need to be increased for Zephyr application to work. The usual thing for Zephyr.

xakep-amatop replied:
So, why did you perform a check in the case when you have exactly the needed number of descriptors? You should notify someone only if you don't have enough, as per my understanding.

Let's imagine a situation where someone transfers always one 512-byte block and allocates only 1 ADMA2 descriptor. It is expected for this person, so why should they see an annoying message in the console?

xakep-amatop commented on drivers/sdhc/sdhci_common.h:
```diff
@@ -225,6 +227,29 @@
+#define SDHC_ADMA_DESC_ADDR_SHIFT 32
+#define SDHC_ADMA_DESC_ADDR_MASK  GENMASK(63, 32)
+
+#define SDHC_ADMA_DESC_MAX_LEN 65532
```
just for curiosity, have you any ideas why so strange number?

-----
I took the liberty of finding the information that interests me (`SD Host Controller Simplified Specification Version 2.00`):
![Screenshot from 2024-06-28 08-01-40](https://github.com/xen-troops/zephyr/assets/27217224/bb14aec7-eee9-43ea-8058-ab4255aaad6f)

GrygiriiS replied:
It's done as in u-boot. The reason is simple - to encoding 65536 byte should be handled as special case by writing 0 in length field. To avoid unnecessary if the desc data size is fixed to  65532.  The Linux Kernel fixes it to 32K also.

xakep-amatop replied:
> It's done as in u-boot. 

So, if you see someone jumping into a well, does that mean you should jump in after them?..

> To encode 65536 bytes, it should be handled as a special case by writing 0 in the length field. To avoid unnecessary checks, the descriptor data size is fixed to 65532 bytes.

This isn't true:
You have a 16-bit `len` field in an ADMA2 descriptor structure. What will happen if you write 65536 to it? What will actually be stored in the `len` field?

GrygiriiS replied:
From my experience, relatively young engineers sometimes think that everything which is done already is ineffective, not optimal and need to be rewritten, trying to get max from HW and use extreme "corner" settings.

By doing so they don't even realize that they are actually the jumping into a well by themselves :)

Yes, I checked what other people/projects have done and I saw nobody is using Zero length in ADMA2 desc. Ok it's weird - but my experience taught me to trust collective experience - if nobody using that smth is wrong.
And I, honestly, have had no desire to investigate it deeper.
 But hence You raised this question and in a bit impolite manner, I spent time on investigation.
 
 So my intuition and experience have not failed me and, of course, I found something:
 
 ```
       if (host->flags & SDHCI_USE_ADMA) {
-               if (host->quirks & SDHCI_QUIRK_BROKEN_ADMA_ZEROLEN_DESC)
+               if (host->quirks & SDHCI_QUIRK_BROKEN_ADMA_ZEROLEN_DESC) {
+                       host->max_adma = 65532; /* 32-bit alignment */
                        mmc->max_seg_size = 65535;
``` 

So, Zero length may be defined by Standard, but not all HW support it, so the way it's done here now is most correct and safest.

xakep-amatop replied:
The value is a little confusing to me without any explanation in the code as to why it was chosen. It's not a good idea to hardcode a strange value for everyone.

> So, zero length may be defined by the standard, but not all hardware supports it. The way it's currently implemented here is the most correct and safest.

I propose creating a configuration option so people can choose. You can define your own values for RPI5 and specify the limits in the configuration option. By default, choose something safe, for example, 32K, since you mentioned that this value is used by the Linux kernel.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+	ctrl = sys_read8(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL);
+
+	if (sdhci_ctx->dma_mode == SDHC_DMA_SDMA) {
+		ctrl |= SDHCI_CTRL_SDMA << SDHCI_CTRL_DMA_SHIFT;
```
What if the controller has the `SDHCI_CTRL_ADMA2` bit set, or `SDHCI_CTRL_SDMA` for the below case?
Should you ensure that the `SDHCI_CTRL_ADMA2` (or `SDHCI_CTRL_SDMA` for the below case) bit isn't set (clear it)?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -687,14 +757,25 @@ static int sdhci_check_version(struct sdhci_common *sdhci_ctx)
+}
+
+int sdhci_enable_adma2(struct sdhci_common *sdhci_ctx, struct sdhc_adma_desc *descs_table,
+		       uint32_t descs_table_szie)
```
here and below: descs_table_s~zi~e ---> descs_table_s**iz**e

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.h:
```diff
@@ -56,6 +56,7 @@
 #define SDHCI_HOST_CONTROL2   0x3E /* 16bit */
 #define SDHCI_CAPABILITIES    0x40 /* 32bit */
 #define SDHCI_CAPABILITIES_1  0x44 /* 32bit */
+#define SDHCI_ADMA_ADDRESS    0x58 /* 32bit */
```
Actually, this register is 64-bit, but the ADMA2 controller can operate in both 32-bit and 64-bit modes, and you can select either mode as far as I understand. Of course, for this to work, the view of `sdhc_adma_desc` should be changed too.

Look at the `Host Control Register (0x28)`, bits 3-4 and `Capabilities Register (0x40)`, bit 28. 

I use the `SD Host Controller Simplified Specification Version 2.00` document for reference.

-----
I believe you should add support for 64-bit mode if it possible by your host controller caps because, in the case of the RPI5, the DMA address can be easily larger than 32 bits.

GrygiriiS replied:
As i mentioned the code is limited to v3.0 and 64-bit dma support not implemented and even removed from v3.0 spec, so it's not going to be implemented. At least not as part of this submission.

xakep-amatop replied:
Again, this isn't true. Only the description of the 64-bit descriptor was removed and described in spec v4 for specs below v4, and according to my information, it was only removed from the **simplified** specification, I don't know if it is  present in a full spec or not.

GrygiriiS replied:
Again I follow what I have - no 64-bit support in "SD Host Controller Simplified Specification Version 3.00".
If needed - anybody can add it if they need.

xakep-amatop replied:
Again, please read messages here https://github.com/xen-troops/zephyr/pull/120/files#r1666383289

xakep-amatop commented on drivers/sdhc/sdhci_common.h:
```diff
@@ -225,6 +227,29 @@
+#define SDHC_ADMA_DESC_ACT1       BIT(4)
+#define SDHC_ADMA_DESC_ACT2       BIT(5)
+#define SDHC_ADMA_DESC_ACT_NOP    0
+#define SDHC_ADMA_DESC_ACT_RSV    SDHC_ADMA_DESC_ACT1
```
I think this define name isn't good because this combination of ACT bits represents RSR only for ADMA2. For ADMA1, it has a different meaning.

GrygiriiS replied:
Only v3.0

xakep-amatop replied:
So, if you don't want to support ADMA1 (spec v3 says it is not required to support ADMA1 by the driver), then, as per my understanding, you should change the naming of the defines here to indicate that these are ADMA2 bits and have nothing related to ADMA1.

GrygiriiS replied:
renamed

xakep-amatop commented on drivers/sdhc/sdhci_common.h:
```diff
@@ -225,6 +227,29 @@
+#define SDHC_ADMA_DESC_ACT_RSV    SDHC_ADMA_DESC_ACT1
+#define SDHC_ADMA_DESC_ACT_TRAN   SDHC_ADMA_DESC_ACT2
+#define SDHC_ADMA_DESC_ACT_LINK   SDHC_ADMA_DESC_ACT2 | SDHC_ADMA_DESC_ACT1
+#define SDHC_ADMA_DESC_LEN_SHIFT  16
```
As far as I understand, this and the defines below should be used instead of the `sdhc_adma_desc` structure. Why do we need both the structure and the shifts? Can we leave only one of them?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -437,7 +503,11 @@ static void sdhci_prepare_dma(struct sdhci_common *sdhci_ctx, struct sdhc_data *
 
 	dma_addr = z_mem_phys_addr(data->data);
```
Should you add a check that ensures `dma_addr` is less than 32-bit, and if it isn't, use non-DMA transfer?
in case of ADMA2 we may try to use 64-bit descs mode

GrygiriiS replied:
fixed. Added sdhci_dma_check() to verify DMA xfer

xakep-amatop commented on drivers/sdhc/Kconfig.brcm:
```diff
@@ -21,14 +21,22 @@ config SDHC_BUFFER_ALIGNMENT
 config BRCM_BCM2712_XFER_SDMA
 	bool "BCM2712 SDHC SDMA data transfer mode"
 
+config BRCM_BCM2712_XFER_ADMA
```
Why isn't the name `BRCM_BCM2712_XFER_ADMA2`?
I don't see that you are trying to enable modes other than ADMA2.

Also, in the case where you have ADMA2 32-bit descriptor mode, the alignment (`SDHC_BUFFER_ALIGNMENT`) can be 32 bits.

GrygiriiS replied:
> Also, in the case where you have ADMA2 32-bit descriptor mode, the alignment (SDHC_BUFFER_ALIGNMENT) can be 32 bits.

Hence, We do not have documentation on RPI5 SoC and taking into account testing  I've done I will stick to the safe alignment which is eq cache line.

GrygiriiS replied:
> Why isn't the name BRCM_BCM2712_XFER_ADMA2?

fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
+				   (sdhci_ctx->adma_descs_num * sizeof(struct sdhc_adma_desc)));
+#endif
+
+	sys_write32((uintptr_t)sdhci_ctx->adma_descs, sdhci_ctx->reg_base + SDHCI_ADMA_ADDRESS);
```
Shouldn't you use the PHYS address here for DMA descriptors?
Of course, if you have direct mapping of virtual to physical addresses, you shouldn't, but is that the case?

GrygiriiS replied:
right. fixed

xakep-amatop reviewed (changes requested):
The review is complete.
I believe I have identified a few significant issues.
Please fix them or correct me if I am mistaken.

xakep-amatop commented on drivers/sdhc/Kconfig.brcm:
```diff
@@ -21,14 +21,22 @@ config SDHC_BUFFER_ALIGNMENT
 
 choice BRCM_BCM2712_XFER_MODE
 	bool "Select BCM2712 SDHC (SD/MMC) data transfer mode"
-	default BRCM_BCM2712_XFER_SDMA
```
Let's discuss this change in more depth. I think with the current implementation of ADMA2, there isn't any compelling reason to use it.

Here are the main differences I see compared to SDMA mode:
- ADMA2 may support 64-bit addresses, but SDMA might also support them. I haven't investigated this question thoroughly; I only know that the driver currently works only with 32-bit addresses.
- ADMA2 can transfer more data than the `block count` register can hold, but this functionality hasn't been implemented.
- **_The main feature, in my opinion, is the ability to link descriptors and use non-linear physical memory chunks._** However, this hasn't been implemented (I even don't know if it is needed for Zephyr), and you are essentially wasting CPU time filling ADMA2 descriptors to function like SDMA.

GrygiriiS replied:
Please pay attention that this implementation supports only **" * - Only SDHC Specification Version 3.0"** which is clearly mentioned in commits description and in sdhci_common.h. Therefore it supports only ADMA2 32-bits. More over, in `"SD Host Controller Simplified Specification Version 3.00"` we have following note:
```
64-bit Descriptor is removed from the Simplified Specification because 64-bit system supported will be
modified by the Host Controller Specification Version 4.00.
```

Next:

1. Making transfers `> then block count register can hold` is, most probably, not a case for Zephyr  (it's ~4G transfer :)). 
2. The scatter/gather in different chanks is also not the case for current Zephyr (at least not smth. supported by current subsystems which are using SD/MMC.
3. The most important part - I'm not going to add features which is impossible to test with HW I do have. Any enhancements are possible for sure, as subsequent PRs in Zephyr mainline (when and if it will go there). 

The main diff between SDMA and ADMA2 is that SDMA has limitation on uninterrupted transfer size which is 512K now, while ADMA2 doesn't have it (taking into account ~4G max transfer size)

xakep-amatop replied:
> Therefore, it supports only ADMA2 32-bit descriptors.

This statement is not true.

> The 64-bit descriptor is removed from the Simplified Specification because 64-bit system support will be modified by the Host Controller Specification Version 4.00.

This means that the 64-bit **descriptor** is not described in the **simplified** documentation. I suppose you don't have the full specification. Even in the simplified spec, there is a lot of information about the support of 64-bit addresses.

Finally, if you have time to open the spec version 4.00, you will find the following:

```
Support of 64-bit addressing is modified from Version 4.00. ADMA2 and ADMA3 64-bit descriptor length is modified to 128 bits considering byte enable alignment. 32-bit or 64-bit addressing mode is selected at initialization, which is determined by the OS. (Up to Version 3.00, 64-bit mode may be selected by the Host Driver at each operation by DMA Select in Host Control 1 register).
```

So, the descriptor format was changed; before v4, in the case of 64-bit addressing, we had 96 bits:
![Screenshot from 2024-07-05 09-10-32](https://github.com/xen-troops/zephyr/assets/27217224/c0bbbce6-0344-43c5-a6db-2af7c22950a9)

> Making transfers larger than the block count register can hold is, most probably, not a case for Zephyr (it's ~4G transfer :)).

First of all, you are making generic code. I don't understand how you calculated 4G. The block count register contains 16 bits according to spec v3, so you can have a maximum of 65535 blocks. The maximum possible block size is 2048 bytes according to the spec, so 65535 * 2048 bytes is approximately 128 Mbytes. Please explain how you arrived at 4G.

Finally, you don't know if you might have a case of a 1-byte block size, which is possible according to the SD Host Control specification.

> The main difference between SDMA and ADMA2 is that SDMA has a limitation on uninterrupted transfer size, which is 512K now, while ADMA2 doesn't have it (taking into account ~4G max transfer size).

This statement is not true either.

Your ADMA2 implementation has similar limitations. Let's calculate: the maximum size of data in one descriptor is 2^16 bytes, and you have only 2^4 descriptors. So, in total, you will have 2^20 bytes, which is 1 Mbyte. To achieve this, you need to fill 16 descriptors. In the case of SDMA, you need to fill the address register two times. I'm not trying to say that you should avoid using ADMA2; this is just for your information.

To use ADMA2 for transferring 4 Gbytes, you need to have 2^16 descriptors, i.e., 0.5 Mbytes per descriptor array (in the case of 32-bit addresses). This is not a problem as per my understanding; it is just for clarification.

GrygiriiS replied:
Ok. if it is sooo important to you - you are right max is 128M and not 4G my mistake. sorry

The 64-bit ADMA2 transfer is not supported by "SD Host Controller **Simplified** Specification Version 3.00" and it's as the fact  there are no required control bits defined in relevant registers (Like Host Control1->DMA Select).
May be somebody has fulls version and need it - This person can easily create and send patch to extend this implementation.

For my use case 32-bit is enough right now and perfectly working.

Above you wrote a lot of word, but what do you propose exactly is probably hidden somewhere deep inside and I failed to see it.

xakep-amatop replied:
> The 64-bit ADMA2 transfer is not supported by "SD Host Controller Simplified Specification Version 3.00" 

Please reread the comment above, and you will find that 64-bit mode is possible.

> it's as the fact there are no required control bits defined in relevant registers (Like Host Control1->DMA Select).
May be somebody has fulls version and need it - This person can easily create and send patch to extend this implementation.

You don't need a full spec; you see that spec 3.0 says the info is moved to spec 4.0. Have you tried to open spec 4.0 or later?

![Screenshot from 2024-07-10 08-49-38](https://github.com/xen-troops/zephyr/assets/27217224/355ab219-25b0-474d-927b-fc2777ff86af)

Some info from spec v3.0:
![Screenshot from 2024-07-10 08-45-18](https://github.com/xen-troops/zephyr/assets/27217224/6923d86a-a874-4dfe-9fd3-68fe5abb1f11)

There are many mentions of support for 64-bit mode in spec v3.0.

> For my use case 32-bit is enough right now and perfectly working.

Hmmm, it will be easy to get a 64-bit addresses when working under Xen. You don't know which pool of addresses will be allocated for the domain, and of course, there are versions of the RPi5 with more than 4GB of RAM.

If we talk more broadly, you are writing a common part of the driver that can be used not only by RPI5 boards, so it is possible that other boards will have addresses above 32 bits.

> Above you wrote a lot of word, but what do you propose exactly is probably hidden somewhere deep inside and I failed to see it.

As it was before, add support for 64-bit mode. **Of course, you could just open an issue for that.** I'm not insisting on fixing it right now; it would be enough to check the address and, in case of a 64-bit address, just redirect it to non-DMA mode. But here, you are just saying that it isn't supported, and your statement is really strange if we take into account the direct speech from the standard in my comments.

I think we need a referee here. @lorc FYI.

GrygiriiS replied:
I completely agree with you. It's agreed that 64-bit ADMA mode will not be implemented.

GrygiriiS reviewed (commented):
Thank you for your review

xakep-amatop reviewed (changes requested):
With this review, I only responded to some of your comments, and some of my comments were left without any update or response. Therefore, I haven't reviewed the new changes.

Please trigger me when you do so.

GrygiriiS reviewed (commented):
Also https://github.com/xen-troops/zephyr/pull/120#discussion_r1666405044
fixed.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -474,6 +545,45 @@ static void sdhci_prepare_data(struct sdhci_common *sdhci_ctx, struct sdhc_comma
+	}
+
+#if defined(CONFIG_64BIT)
+	/* Only 32bit DMA supported */
```
Why can't you use the buffer register, aka non-DMA mode? Do we really need to panic here?

xakep-amatop replied:
discussed, an user should enable non-DMA mode by config it they want to work with 64bit addresses

GrygiriiS replied:
I consider as critical mis-configuration with current implementation, so yes panic or assert

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -474,6 +545,45 @@ static void sdhci_prepare_data(struct sdhci_common *sdhci_ctx, struct sdhc_comma
+
+#if defined(CONFIG_64BIT)
+	/* Only 32bit DMA supported */
+	if (((uint64_t)data->data >> 32)) {
```
Is it possible that we have a 32-bit virtual address and a 64-bit DMA address for it?
Shouldn't you check the DMA address here instead of the virtual one?

GrygiriiS replied:
fixed

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -474,6 +545,45 @@ static void sdhci_prepare_data(struct sdhci_common *sdhci_ctx, struct sdhc_comma
+	if (sdhci_ctx->dma_mode == SDHC_DMA_SDMA) {
+		if (trans_bytes > SDHC_SDMA_XFER_MAX) {
+			LOG_ERR("sdhc:sdma: max xfer size overflow max:%08d", SDHC_SDMA_XFER_MAX);
+			return -EIO;
```
Note: in this and the following cases, we can continue to transfer in non-DMA mode.

xakep-amatop replied:
discussed, an user should enable non-DMA mode by configuring it if they want to work with 64-bit addresses.

GrygiriiS replied:
XFER mode is selected during init and dynamic change is not supported now. I have no plans adding such feature now.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -695,14 +809,32 @@ static int sdhci_check_version(struct sdhci_common *sdhci_ctx)
 
+#if defined(CONFIG_64BIT)
+	/* Only 32bit DMA supported */
+	if (((uint64_t)descs_table >> 32)) {
```
Is it possible that the DMA address is 64 bits and the virtual address is 32 bits for this descriptor table? If yes, could you please check the DMA address here instead of the virtual one?

And isn't it enough to add some print and to disable ADMA2 support without panicking ?

xakep-amatop replied:
> And isn't it enough to add some print and to disable ADMA2 support without panicking ?

discussed, an user should enable non-DMA mode by configuring it if they want to work with 64-bit addresses.

GrygiriiS replied:
It's very unsafe and not smth expected by Zephyr where max features configured during init. User can switch for PIO mode now for 64bit xfer or anybody who need can add 64bit implementation in the future.

xakep-amatop reviewed (commented):
The review has been finished.
Some of old conversations weren't completed, and I have a few questions related to the new changes.

Lets discuss.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -695,14 +809,32 @@ static int sdhci_check_version(struct sdhci_common *sdhci_ctx)
+		return -EINVAL;
 	}
 
+#if defined(CONFIG_64BIT)
```
I don't know exactly how it works, but as far as I understand, it is possible to have addresses bigger than 32 bits on a 32-bit system. This is known as PAE for x86 and LPAE for ARM.

GrygiriiS replied:
Those >32bit addrs usually are not supported for DMA - at least I'm not aware of  such cases. In such, systems DMA is allowed only in <32bit area.

GrygiriiS commented:
changed SDHCI_ADMA_ADDRESS to SDHCI_ADMA_ADDRESS_LOW
added check for max block cnt 65535
changed check for 64-bit addr to used dma_addr
added CONFIG_SDHCI_GENERIC_ADMA2_DESC_MAX_LEN

xakep-amatop commented on drivers/sdhc/Kconfig.sdhc:
```diff
@@ -7,3 +7,11 @@ config SDHCI_GENERIC
+
+config SDHCI_GENERIC_ADMA2_DESC_MAX_LEN
+	int "ADMA2 max xfer size per descriptor"
+	default 65532
```
You can limit the minimum and maximum number for an integer option by adding the `range` property:
`range 1 65536`

GrygiriiS replied:
fixed

xakep-amatop reviewed (approved):
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`

oleksiimoisieiev commented:
LGTM `Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
